doc/ctpv.1
'\" t
.ds ub \(:Uberzug
.
.TH CTPV 1 2022 Linux "User's Reference Manuals"
.
.SH NAME
ctpv \- terminal previewer
.
.SH SYNOPSIS
.
.SY ctpv
.RI [ \-d ]
.I  file
.RI [ w ]
.RI [ h ]
.RI [ x ]
.RI [ y ]
.RI [ id ]
.YS
.
.SY ctpv
.B \-h
.YS
.
.SY ctpv
.B \-l
.YS
.
.SY ctpv
.B \-m
.I file
\&.\|.\|.\&
.YS
.
.SY ctpv
.RB { \-s | \-c | \-e }
.I id
.YS
.
.SH DESCRIPTION
.
.B ctpv
is a previewer utility made for integration into other programs like
.IR lf .
.
.PP
When
.B ctpv
is given a
.IR file ,
it determines an appropriate preview for it and runs it.
Depending on the preview and installed programs, it can print
some text to standard output or display an image.
Symbolic links are dereferenced.
.
.SS Built-in previews and dependencies
.
Previewing each file type requires specific programs.
If a program is not found on the system,
.B ctpv
will try to use another one.
Only one program is required for each file type.
For example, you only need either
.IR elinks ,
.I lynx
or
.I w3m
installed on your system to view HTML files.
.
.PP
.
.\" This table is auto generated!
.
.\" TABLESTART
.TS
allbox;
lb lb
l li .
File type	Programs
any	T{
exiftool cat
T}
archive	T{
atool
T}
audio	T{
ffmpegthumbnailer ffmpeg
T}
diff	T{
colordiff delta diff\-so\-fancy
T}
directory	T{
ls
T}
font	T{
fontimage
T}
gpg-encrypted	T{
gpg
T}
html	T{
elinks lynx w3m
T}
image	T{
ueberzug chafa
T}
json	T{
jq
T}
markdown	T{
glow mdcat
T}
office	T{
libreoffice
T}
pdf	T{
pdftoppm
T}
svg	T{
convert
T}
text	T{
bat cat highlight source\-highlight
T}
torrent	T{
transmission\-show
T}
video	T{
ffmpegthumbnailer
T}
.TE
.\" TABLEEND
.
.SS Integration
.
.TP
.IR lf \~\c
file manager
Add this snippet to
.I lf
configuration file (usually located at
.IR \(ti/.config/lf/lfrc ).
.
.RS
.IP
.EX
set previewer ctpv
set cleaner ctpvclear
&ctpv \-s $id
&ctpvquit $id
.EE
.RE
.
.SS Image previews
.
Image previews are enabled by either installing
.I \*(ub
or
.I Chafa
(as seen in
.IR "Built-in previews and dependencies" )
or using built-in image preview functionality of
.I Kitty
terminal.
.
.PP
When possible,
.B ctpv
will prefer
.I \*(ub
over others image preview methods. You can override this
behavior by using
.B forcekitty
or
.B forcechafa
options (see
.IR "Preview options" ).
.
.PP
.I \*(ub
supports
.I X11
only, so
.I Wayland
is not supported.
To enable high-resolution image previews on
.IR Wayland ,
you need to follow these steps:
.
.IP \(bu 4
Make sure you use one of the
.UR https://\:www\:.arewesixelyet\:.com
terminals that support sixel
.UE
.IP \(bu 4
Install
.UR https://\:github\:.com/\:horriblename/\:lf
.I lf-sixel
.UE .
.
.IP \(bu
Install
.I Chafa
.
.IP \(bu
Add
.B chafasixel
configuration option (see
.IR "Preview options" )
.
.PP
The fork is required because, as of 2023-03-19,
the original lf does not support sixel protocol.
.
.SS How previews are selected
.
Initially,
.B ctpv
retrieves MIME type and extension from
.I file
passed as the first argument (MIME type is extracted using
.BR libmagic (3)).
.
.PP
Then it creates a list of all previews respecting user
configuration in a special order, where previews that are
more specific appear at the top and more generic ones at the bottom.
The list can be viewed by using
.B \-l
option. The order can be changed (see
.IR "Setting priority" ).
.
.PP
Finally,
.B ctpv
goes through the list starting with the first element
and checks if a preview matches
.IR file 's
extension and MIME type.
If it does, it runs a preview script.
If the script exits with status 127
(which usually means that a program that is necessary for generating
a preview content is not installed on the system),
.B ctpv
attempts to run another appropriate preview and so on.
If the script exists with 0 or a any other status, both standard output
and standard error are printed.
.
.SH OPTIONS
.
.TP
.B \-h
Print help message.
.
.TP
.B \-l
List all previews.
.
.TP
.BR \-m \~\c
.IR file \~.\|.\|.\&
Print extension and MIME type of
.IR file .
.
.TP
.BR \-s \~\c
.I id
Start server with ID
.IR id .
.
.TP
.BR \-c \~\c
.I id
Send
.B clear
command to server with ID
.I id
(usually, it removes image from terminal).
.
.TP
.BR \-e \~\c
.I id
Kill server with ID
.IR id .
.
.TP
.B \-d
Enable debug messages.
.
.SH CONFIGURATION
.
.B ctpv
uses a configuration file usually located at
.I \(ti/.config/ctpv/config
(see
.IR FILES ).
Its format somewhat resembles one used by
.IR lf .
There are several commands that can be used to add
previews or set different settings.
Commands are separated by newlines.
Comments start with number sign
.RB \(oq # \(cq.
.
.PP
Example:
.
.IP
.EX
# Set some options
set forcekitty
set shell "/usr/bin/bash"
.sp
# Add a new preview
preview cow .moo {{
\&    cowsay < "$f"
}}
.sp
# Remove some previews
remove w3m
remove lynx
remove elinks
.EE
.
.SS Preview options
.
An option can be set using
.B set
command.
.
.TP
.BR shell \~\c
.RI \(dq path \(dq
Use
.I path
as a path to a shell to run previews with.
Use it if you have a non-POSIX compliant shell installed as a default shell.
The setting defaults to
.BR /bin/sh .
.
.TP
.B forcekitty
Always use
.I Kitty
terminal's built-in method of previewing images.
.
.TP
.B forcekittyanim
Always use
.I Kitty
terminal's built-in method of previewing images for animated
images.
.
.TP
.B forcechafa
Always use
.I Chafa
for image previews.
.
.TP
.B noimages
Print only text and do not use any image previewing method.
.
.TP
.B nosymlinkinfo
Do not print resolved path of symbolic links.
.
.TP
.B chafasixel
Set output format of
.I Chafa
to \(lqsixels\(rq instead of \(lqsymbols\(rq.
Use it if your file manager and terminal are capable of properly displaying
sixel data.
.
.TP
.B showgpg
Preview
.BR gpg (1)
encrypted files.
Filename must have \(lq.gpg\(rq extension.
.
.SS Defining custom previews
.
User-defined previews are added with
.B preview
command.
.
.PP
An example below defines a new preview with name \(lqmanpage\(rq
that applies to files with extension \(lq.1\(rq.
A preview itself is a shell script enclosed within double curly
braces.
.
.IP
.EX
preview manpage .1 {{
\&    groff \-man \-tep \-Tutf8 \-rLL="${w}n" "${f}" | col \-x
}}
.EE
.
.PP
Running
.I "ctpv\~file.1"
where
.I file.1
is a source code for a manpage will run
.BR groff (1)
to produce a formatted manpage like the one you are reading.
.
.PP
Manpages filenames may also end with other extensions:
\(lq.2\(rq, \(lq.3\(rq, \(lq.4\(rq and so on.
It's possible to make user-defines previews apply to several
file types at once:
.
.IP
.EX
preview manpage .1 .2 .3 .4 .5 .6 .7 .8 {{
\&    # groff command
}}
.EE
.
.PP
Variable
.B $f
stores
.I file
that was passed as a first argument to
.BR ctpv .
It's strongly suggested to enclose
.B $f
with double quotes
.RB ( \(dq$f\(dq )
because otherwise the script will not work as
expected if
.B $f
stores a filename with whitespace.
.
.PP
There are other variables that are exported into preview
script environment:
.BR $w ,
.BR $h ,
.BR $x ,
.B $y
and
.BR $id .
There are also
.B $m
and
.B $e
which store MIME type and extension of
.IR file .
.
.PP
You can specify MIME type instead of filename extension
in preview definition:
.
.IP
.EX
preview json_example application/json {{
\&    # preview json files
}}
.EE
.
.PP
You can omit subtype part of the MIME type
by replacing it with
.RB \(oq * \(cq.
.
.IP
.EX
preview any_text_example text/* {{
\&    # this one applies to all text files
}}
.EE
.
.PP
Setting subtype to
.RB \(oq * \(cq
will make the preview above apply to any file which MIME type starts with
.BR text/ .
.
.SS Setting priority
.
If there are several previews that apply to the same file type,
only the top one in the list is chosen (see
.IR "How previews are selected" ).
To alter this behavior, you can use
.B priority
command to change preview priority:
.
.IP
.EX
priority cat
.EE
.
.PP
The snippet above sets priority of a built-in preview named \(lqcat\(rq
to 1, thus now it's used for all text files.
It's possible to specify an integer as the second argument
to set priority other than 1 (may also be negative).
.
.SS Removing previews
.
.B remove
command simply removes a preview (also works for built-in ones):
.
.IP
.EX
remove cat
.EE
.
.PP
It's useful if you have a program installed on your system but you
don't want
.B ctpv
to use it for generating previews.
.
.SH ENVIRONMENT
.
.TP
.I id
.I id
of a server to connect to
(see
.B \-s
option).
.
.SH FILES
.
.TP
.I $XDG_CONFIG_HOME/ctpv/config
Configuration file.
If
.I $XDG_CONFIG_HOME
is not set, defaults to
.IR \(ti/.config .
.
.TP
.I $XDG_CACHE_HOME/ctpv
Directory to store cached image previews.
It takes some time to generate an image preview for some file types,
such as videos or PDF files, this is why the generated images are
stored in the directory to be shown if the same file is previewed
again.
If
.I $XDG_CACHE_HOME
is not set, defaults to
.IR \(ti/.cache .
.
.SH SEE ALSO
.
.BR lf (1)
.
.SH AUTHOR
.
Written by Nikita Ivanov.


src/ctpv.c
#include <stdio.h>
#include <magic.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>
#include <sys/stat.h>
#include <openssl/md5.h>

#include "ctpv.h"
#include "error.h"
#include "utils.h"
#include "config.h"
#include "server.h"
#include "preview.h"
#include "../version.h"
#include "../previews.h"
#include "../gen/help.h"

struct InputFile {
    char link[PATH_MAX], path[PATH_MAX];
};

struct CTPV ctpv;

const char any_type[] = ANY_TYPE;

static magic_t magic;

static Parser *parser;

static VectorPreview *previews;

static void cleanup(void)
{
    previews_cleanup();
    if (parser)
        config_cleanup(parser);
    if (magic != NULL)
        magic_close(magic);
    if (previews)
        vectorPreview_free(previews);
}

static RESULT init_magic(void)
{
    ERRCHK_RET_MSG(!(magic = magic_open(MAGIC_MIME_TYPE)), magic_error(magic));

    ERRCHK_RET_MSG(magic_load(magic, NULL) != 0, magic_error(magic));

    return OK;
}

static RESULT create_dir(char *buf, size_t len)
{
    char dir[len];
    strncpy(dir, buf, LEN(dir) - 1);
    ERRCHK_RET_ERN(mkpath(dir, 0700) == -1);

    return OK;
}

static RESULT get_config_file(char *buf, size_t len)
{
    ERRCHK_RET_OK(get_config_dir(buf, len, "ctpv/"));
    ERRCHK_RET_OK(create_dir(buf, len));

    strncat(buf, "config", len - 1);

    if (access(buf, F_OK) != 0)
        close(creat(buf, 0600));

    return OK;
}

static RESULT config(int prevs)
{
    char config_file[FILENAME_MAX];
    ERRCHK_RET_OK(get_config_file(config_file, LEN(config_file)));

    ERRCHK_RET_OK(config_load(&parser, prevs ? previews : NULL, config_file));

    return OK;
}

static RESULT init_previews(void)
{
    /* 20 is some arbitrary number, it's here in order to
     * to save one realloc() if user has less then 20 custom previews */
    previews = vectorPreview_new(LEN(b_previews) + 20);
    vectorPreview_append_arr(previews, b_previews, LEN(b_previews));

    ERRCHK_RET_OK(config(1));

    previews_init(previews->buf, previews->len);

    return OK;
}

static const char *get_mimetype(const char *path)
{
    const char *r = magic_file(magic, path);
    if (!r) {
        FUNCFAILED("magic_file", magic_error(magic));
        return NULL;
    }

    return r;
}

static inline void file_access_err(char *f, int errno_)
{
    print_errorf("failed to access '%s': %s", f, strerror(errno_));
}

static RESULT get_input_file(struct InputFile *input_f, char *f)
{
    if (!f) {
        print_error("file not given");
        return ERR;
    }

    input_f->link[0] = input_f->path[0] = '\0';

    if (realpath(f, input_f->path) == NULL) {
        file_access_err(input_f->path, errno);
        return ERR;
    }

    ssize_t link_len = readlink(f, input_f->link, LEN(input_f->link));
    if (link_len == -1 && errno != EINVAL) {
        FUNCFAILED("readlink", strerror(errno));
        return ERR;
    }

    input_f->link[link_len] = '\0';

    return OK;
}

static RESULT is_newer(int *resp, char *f1, char *f2)
{
    struct stat stat1, stat2;
    ERRCHK_RET_ERN(lstat(f1, &stat1) == -1);
    ERRCHK_RET_ERN(lstat(f2, &stat2) == -1);

#if __APPLE__
    struct timespec *t1 = &stat1.st_ctimespec;
    struct timespec *t2 = &stat2.st_ctimespec;
#else
    struct timespec *t1 = &stat1.st_ctim;
    struct timespec *t2 = &stat2.st_ctim;
#endif

    time_t sec_d = t1->tv_sec - t2->tv_sec;
    if (sec_d < 0)
        goto older;
    else if (sec_d == 0 && t1->tv_nsec <= t2->tv_nsec)
        goto older;

    *resp = 1;
    return OK;

older:
    *resp = 0;
    return OK;
}

static void md5_string(char *buf, size_t len, char *s)
{
    unsigned char out[MD5_DIGEST_LENGTH];
    char b[16];

    MD5((const unsigned char *)s, strlen(s), out);

    buf[0] = '\0';
    for (unsigned int i = 0; i < LEN(out); i++) {
        snprintf(b, LEN(b)-1, "%02x", out[i]);
        strncat(buf, b, len);
    }
}

static RESULT get_cache_file(char *dir, size_t dir_len, char *filename,
                             size_t filename_len, char *file)
{
    ERRCHK_RET_OK(get_cache_dir(dir, dir_len, "ctpv/"));
    ERRCHK_RET_OK(create_dir(dir, dir_len));

    size_t dir_str_len = strlen(dir);

    memcpy(filename, dir, filename_len);

    md5_string(filename + dir_str_len, filename_len - dir_str_len - 1, file);

    /* Remove dash at the end */
    dir[dir_str_len-1] = '\0';

    return OK;
}

static RESULT check_cache(int *resp, char *file, char *cache_file)
{
    if (access(cache_file, F_OK) != 0) {
        *resp = 0;
        return OK;
    }

    return is_newer(resp, cache_file, file);
}

#define GET_PARG(a, i) (a) = (argc > (i) ? argv[i] : NULL)

static RESULT preview(int argc, char *argv[])
{
    char *f, *w, *h, *x, *y, *id;
    char w_buf[24], h_buf[24], y_buf[24];
    long w_l, h_l, y_l;

    GET_PARG(f, 0);
    GET_PARG(w, 1);
    GET_PARG(h, 2);
    GET_PARG(x, 3);
    GET_PARG(y, 4);
    GET_PARG(id, 5);

    if (!w)
        w = "80";
    if (!h)
        h = "40";
    if (!x)
        x = "0";
    if (!y)
        y = "0";

    ERRCHK_RET_OK(strtol_w(&w_l, w, NULL, 10));
    ERRCHK_RET_OK(strtol_w(&h_l, h, NULL, 10));
    ERRCHK_RET_OK(strtol_w(&y_l, y, NULL, 10));

    ERRCHK_RET_OK(init_previews());

    struct InputFile input_f;
    ERRCHK_RET_OK(get_input_file(&input_f, f));

    if (!ctpv.opts.nosymlinkinfo && *input_f.link) {
        printf("\033[1;36mSymlink points to:\033[m\n\t%s\n\n", input_f.link);
        fflush(stdout);

        y_l += 3;
        h_l -= 3;
    }

    /* To some reason lf chops off last 2 characters of a line */
    w_l = MAX(0, w_l - 2);

    snprintf(w_buf, LEN(w_buf), "%ld", w_l);
    snprintf(h_buf, LEN(h_buf), "%ld", h_l);
    snprintf(y_buf, LEN(y_buf), "%ld", y_l);

    w = w_buf;
    h = h_buf;
    y = y_buf;

    ERRCHK_RET_OK(init_magic());

    const char *mimetype;
    ERRCHK_RET(!(mimetype = get_mimetype(input_f.path)));

    char cache_dir[FILENAME_MAX], cache_file[FILENAME_MAX];
    ERRCHK_RET_OK(get_cache_file(cache_dir, LEN(cache_file), cache_file,
                                 LEN(cache_file), input_f.path));

    int cache_valid;
    ERRCHK_RET_OK(check_cache(&cache_valid, input_f.path, cache_file));

    PreviewArgs args = {
        .f = input_f.path,
        .w = w,
        .h = h,
        .x = x,
        .y = y,
        .id = id,
        .cache_dir = cache_dir,
        .cache_file = cache_file,
        .cache_valid = cache_valid,
    };

    return preview_run(get_ext(input_f.path), mimetype, &args);
}

static RESULT server(void)
{
    return server_listen(ctpv.server_id_s);
}

static RESULT clear(void)
{
    ERRCHK_RET_OK(config(0));
    return server_clear(ctpv.server_id_s);
}

static RESULT end(void)
{
    ERRCHK_RET_OK(config(0));
    return server_end(ctpv.server_id_s);
}

static RESULT list(void)
{
    ERRCHK_RET_OK(init_previews());

    size_t len;
    Preview p, **list = previews_get(&len);
    const char *n, *e, *t, *s;

    const char header_name[] = "Name", header_ext[] = "Extension",
               header_mime[] = "MIME type";

    int width_name = 0, width_ext = 0;

    for (size_t i = 0; i < len + 1; i++) {
        if (i < len) {
            p = *list[i];
            n = p.name;
            e = p.ext;
        } else {
            n = header_name;
            e = header_ext;
        }

        int name_len = strlennull(n);
        int ext_len = strlennull(e);
        width_name = MAX(width_name, name_len);
        width_ext = MAX(width_ext, ext_len);
    }

    width_name += 2, width_ext += 2;

    puts("List of available previews:\n");
    printf("\t%-*s %-*s %s\n\n", width_name, header_name, width_ext, header_ext,
           header_mime);

    for (size_t i = 0; i < len; i++) {
        p = *list[i];
        e = p.ext;
        t = p.type;
        s = p.subtype;

        if (!e)
            e = any_type;

        if (!t) {
            t = any_type;
            s = any_type;
        } else if (!s) {
            s = any_type;
        }

        printf("\t%-*s .%-*s %s/%s\n", width_name, p.name, width_ext - 1, e, t,
               s);
    }

    puts("\nNote: '" ANY_TYPE "' means that it matches any.\n");

    return OK;
}

static RESULT mime(int argc, char *argv[])
{
    const char *mimetype;
    struct InputFile input_f;

    if (argc <= 0) {
        print_error("files are not specified");
        return ERR;
    }

    ERRCHK_RET_OK(init_magic());

    for (int i = 0; i < argc; i++) {
        ERRCHK_RET_OK(get_input_file(&input_f, argv[i]));

        mimetype = get_mimetype(input_f.path);
        ERRCHK_RET(!mimetype);

        if (argc > 1)
            printf("%s:\t", argv[i]);

        printf(".%s ", get_ext(input_f.path));
        puts(mimetype);
    }

    return OK;
}

static RESULT help(void)
{
    printf("Usage: %s [OPTION]... FILE\n\n", program);
    printf("Options:\n%s", help_txt);
    return OK;
}

static RESULT version(void)
{
    printf("%s version %s\n", program, VERSION);
    return OK;
}

int main(int argc, char *argv[])
{
    program = argc > 0 ? argv[0] : "ctpv";

    ctpv.opts.shell = "/bin/sh";
    ctpv.debug = 0;

    int c;
    while ((c = getopt(argc, argv, "hs:c:e:lmdv")) != -1) {
        switch (c) {
        case 'h':
            ctpv.mode = MODE_HELP;
            break;
        case 's':
            ctpv.mode = MODE_SERVER;
            ctpv.server_id_s = optarg;
            break;
        case 'c':
            ctpv.mode = MODE_CLEAR;
            ctpv.server_id_s = optarg;
            break;
        case 'e':
            ctpv.mode = MODE_END;
            ctpv.server_id_s = optarg;
            break;
        case 'l':
            ctpv.mode = MODE_LIST;
            break;
        case 'm':
            ctpv.mode = MODE_MIME;
            break;
        case 'd':
            ctpv.debug = 1;
            break;
        case 'v':
            ctpv.mode = MODE_VERSION;
            break;
        default:
            return EXIT_FAILURE;
        }
    }

    argc -= optind;
    argv = &argv[optind];

    enum Result ret;
    switch (ctpv.mode) {
    case MODE_PREVIEW:
        ret = preview(argc, argv);
        break;
    case MODE_HELP:
        ret = help();
        break;
    case MODE_SERVER:
        ret = server();
        break;
    case MODE_CLEAR:
        ret = clear();
        break;
    case MODE_END:
        ret = end();
        break;
    case MODE_LIST:
        ret = list();
        break;
    case MODE_MIME:
        ret = mime(argc, argv);
        break;
    case MODE_VERSION:
        ret = version();
        break;
    default:
        PRINTINTERR("unknown mode: %d", ctpv.mode);
        ret = ERR;
        break;
    }

    cleanup();

    return ret == OK ? EXIT_SUCCESS : EXIT_FAILURE;
}
